lib/src/socket.dart

```diff
@@ -392,211 +335,217 @@ class PhoenixSocket {
-      _logger.severe(
-        '[phoenix_socket] Heartbeat message failed: WebSocketChannelException',
-        err,
-        stacktrace,
+    } on WebSocketChannelException catch (error, stackTrace) {
+      _logger.warning(
+        'Heartbeat message failed: WebSocketChannelException',
+        error,
+        stackTrace,
       );
-      _triggerChannelExceptions(PhoenixException(
-        socketError: PhoenixSocketErrorEvent(
-          error: err,
-          stacktrace: stacktrace,
+      _stateEventStreamController.add(
+        PhoenixSocketErrorEvent(
+          error: error,
+          stacktrace: stackTrace,
         ),
-      ));
-
-      return false;
-    } catch (err, stacktrace) {
-      _logger.severe(
-        '[phoenix_socket] Heartbeat message failed',
-        err,
-        stacktrace,
       );
 
+      return false;
+    } catch (error, stackTrace) {
+      _logger.warning('Heartbeat message failed', error, stackTrace);
+      _reconnect(4001, reason: 'Heartbeat timeout');
       return false;
     }
   }
 
-  void _triggerChannelExceptions(PhoenixException exception) {
-    _logger.fine(
-      () => 'Trigger channel exceptions on ${channels.length} channels',
-    );
-    for (final channel in channels.values) {
-      _logger.finer(
-        () => 'Trigger channel exceptions on ${channel.topic}',
+  Timer _scheduleHeartbeat([String? previousHeartbeat]) {
+    return Timer(_options.heartbeat, () {
+      if (previousHeartbeat != null) {
+        final completer = _pendingMessages.remove(previousHeartbeat);
+        if (completer != null && !completer.isCompleted) {
+          completer.completeError(
+            TimeoutException(
+              'Heartbeat $previousHeartbeat not completed before sending new one',
+            ),
+            StackTrace.current,
+          );
+          return;
+        }
+      }
+
+      _sendHeartbeat();
+    });
+  }
+
+  void _cancelHeartbeat() {
+    if (_heartbeatTimeout != null) {
+      _heartbeatTimeout?.cancel();
+      _heartbeatTimeout = null;
+    }
+    if (_latestHeartbeatRef != null) {
+      _pendingMessages.remove(_latestHeartbeatRef)?.complete(
+            Message.heartbeat(_latestHeartbeatRef!), // doesn't matter
+          );
+      _latestHeartbeatRef = null;
+    }
+  }
+
+  bool _shouldPipeMessage(String message) {
+    final currentSocketState = _socketStateStream.valueOrNull;
+    if (currentSocketState is WebSocketReady) {
+      return true;
+    } else {
+      _logger.warning(
+        'Message from socket dropped because PhoenixSocket is not ready (is $currentSocketState)'
+        '\n$message',
       );
-      channel.triggerError(exception);
+
+      return false;
     }
   }
 
-  Message _heartbeatMessage() => Message.heartbeat(_nextHeartbeatRef = nextRef);
+  /// EVENT HANDLERS
+
+  // Exists for testing only
+  void onSocketDataCallback(String message) {
+    if (_shouldPipeMessage(message)) {
+      if (!_receiveStreamController.isClosed) {
```

Comment:
Should this `!_receiveStreamController.isClosed` be inside `_shouldPipeMessage` as well?

Reply:
👍 Not a strong opinion if you think it's clearer. Does it need the nested if statement though 🤔 

```
if (_shouldPipeMessage(message) && !_receiveStreamController.isClosed) {
        _receiveStreamController.add(message);
}
```

Reply:
Done.